src/server/editorServices.ts

```diff
@@ -855,65 +858,247 @@ namespace ts.server {
+                }
+                return configFilePresenceInfo.exists;
+            }
+
+            // Theorotically we should be adding watch for the directory here itself.
+            // In practice there will be very few scenarios where the config file gets added
+            // somewhere inside the another config file directory.
+            // And technically we could handle that case in configFile's directory watcher in some cases
+            // But given that its a rare scenario it seems like too much overhead. (we werent watching those directories earlier either)
+            // So what we are now watching is: configFile if the project is open
+            // And the whole chain of config files only for the inferred project roots
+
+            // Cache the host value of file exists and add the info tio to the tracked root
+            const trackingOpenFilesMap = createMap<boolean>();
+            trackingOpenFilesMap.set(info.path, false);
+            const exists = this.host.fileExists(configFileName);
+            configFilePresenceInfo = { exists, trackingOpenFilesMap };
+            this.mapOfConfigFilePresence.set(canonicalConfigFilePath, configFilePresenceInfo);
+            this.logConfigFileWatchUpdate(configFileName, configFilePresenceInfo, ConfigFileWatcherStatus.TrackingFileAdded);
+            return exists;
+        }
+
+        private setConfigFilePresenceByNewConfiguredProject(project: ConfiguredProject) {
+            const configFilePresenceInfo = this.mapOfConfigFilePresence.get(project.canonicalConfigFilePath);
+            if (configFilePresenceInfo) {
+                Debug.assert(configFilePresenceInfo.exists);
+                // close existing watcher
+                if (configFilePresenceInfo.configFileWatcher) {
+                    const configFileName = project.getConfigFilePath();
+                    this.closeFileWatcher(
+                        WatchType.ConfigFileForInferredRoot, /*project*/ undefined, configFileName,
+                        configFilePresenceInfo.configFileWatcher, WatcherCloseReason.ConfigProjectCreated
+                    );
+                    configFilePresenceInfo.configFileWatcher = undefined;
+                    this.logConfigFileWatchUpdate(configFileName, configFilePresenceInfo, ConfigFileWatcherStatus.UpdatedCallback);
+                }
+            }
+            else {
+                // We could be in this scenario if it is the external project tracked configured file
+                // Since that route doesnt check if the config file is present or not
+                this.mapOfConfigFilePresence.set(project.canonicalConfigFilePath, {
+                    exists: true,
+                    trackingOpenFilesMap: createMap<boolean>()
+                });
+            }
+        }
+
+        private configFileExistenceTracksInferredRoot(configFilePresenceInfo: ConfigFileExistence) {

```

Comment:
Could you document this? It appears to return `true` if at least one file is open. What is an "inferred root"?